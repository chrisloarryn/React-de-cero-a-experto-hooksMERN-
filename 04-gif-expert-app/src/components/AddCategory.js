import React, { useState } from 'react'
import PropTypes from 'prop-types'

export const AddCategory = ({ setCategories }) => {
  const [inputValue, setInputValue] = useState('Hello World')

  const handleInputChange = e => {
    e.preventDefault()
    setInputValue(e.target.value)
    // console.log('handleInputChange invoked')
  }

  const handleSubmit = e => {
    e.preventDefault()
    // console.log('handleSubmit', inputValue)
    setCategories(cats => [{ name: inputValue }, ...cats])
    setInputValue('')
  }

  return (
    <form onSubmit={handleSubmit}>
      <p>{inputValue}</p>
      <input type='text' value={inputValue} onChange={handleInputChange} />
    </form>
  )
}

AddCategory.propTypes = {
  setCategories: PropTypes.func.isRequired
}
