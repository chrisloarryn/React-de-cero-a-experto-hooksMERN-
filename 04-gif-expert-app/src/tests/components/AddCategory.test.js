import React from 'react'
import '@testing-library/jest-dom'

import { shallow } from 'enzyme'
import { AddCategory } from '../../components/AddCategory'

describe(`Tests in <AddCategory /> Component`, () => {
  const setCategories = jest.fn()
  let wrapper = shallow(<AddCategory setCategories={setCategories} />)
  beforeEach( () => {
      jest.clearAllMocks()
      wrapper = shallow(<AddCategory setCategories={setCategories} />)
  })

  test('should be rendered correctly', () => {
    expect(wrapper).toMatchSnapshot()
  })

  test('should change text box', () => {
    const input = wrapper.find('input')
    const value = 'Hello world from test'
    input.simulate('change', {
      target: {
        value
      },
      preventDefault(){}
    })
    expect(
      wrapper
        .find('p')
        .text()
        .trim()
    ).toBe(value)
  })

  test(`should NOT post info with the submit handler`, () => {
    wrapper.find('form').simulate('submit', { preventDefault(){} })
    expect(setCategories).toHaveBeenCalledTimes(1)
  })

  test('should call setCategories and clean the box', () => {
    
    const value = 'test n3'
    wrapper.find('input').simulate('change', {
        target: {
          value
        },
        preventDefault(){}
      })

    // let text = wrapper.find('p').text().trim()

    wrapper.find('form').simulate('submit', { preventDefault(){} })
    
    expect(setCategories).toHaveBeenCalled()
    expect(setCategories).toBeCalledTimes(1)
    expect(setCategories).toHaveBeenLastCalledWith(expect.any(Function))
    
    const input = wrapper.find('input').prop('value')
    expect(input).toBe('')
  });
})
