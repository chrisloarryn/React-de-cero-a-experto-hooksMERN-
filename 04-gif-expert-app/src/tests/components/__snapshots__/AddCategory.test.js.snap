// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in <AddCategory /> Component should be rendered correctly 1`] = `
<form
  onSubmit={[Function]}
>
  <p>
    Hello World
  </p>
  <input
    onChange={[Function]}
    type="text"
    value="Hello World"
  />
</form>
`;
